Feature/2.x/72 module javascript needs improvement

ctorgalson opened the pull request:
<!-- See https://docs.localgovdrupal.org/contributing/ for guidelines on contributing. -->

Closes #72

## What does this change?

TL;DR: this addresses the accessibility, translatability, and theming issues in #72 without altering the themeable markup:

---

**NOTE**: if we we're willing to accept breaking changes (mainly to the markup), this could be rewritten much more efficiently.

---

- all text provided by the js is now passed through `Drupal.t()` (sometimes via theme functions),
- the buttons--including icons where relevant--are now provided by `Drupal.theme` functions that can be overridden by themes,
- the javascript is reasonably compact and much better organised (~65 lines of setup code calling 6 main functions and four `Drupal.theme` functions,
- removes `aria-expanded` attribute from "Show summaries" button,
- adds `aria-controls` attribute to "Show summaries" button,
- the javascript now sets `aria-controls` attributes on summaries, and a corresponding `id` attribute on the individual "Show summaries" buttons,
- the javascript uses `Drupal.announce()` to say "All summaries expanded" or "All summaries collapsed' when the "Show summaries" button is used,

## How to test

<!-- Provide instructions to help others verify the change. This could take the
form of "On main, do X and witness Y. On this branch, do X and witness Z. " -->

- check out this branch of the module into an LGD site
- clear Drupal caches
- view one or more step by step pages
- expand and collapse some summaries
- expand and collapse the "Show summaries" button

## How can we measure success?

With the following exceptions, there should be no differences in behaviour from the current implementation:

- screenreader users will no longer hear the "Show summaries" button's 'expanded' status when it's clicked (`aria-expanded` removed)
- screenreader users will now hear "All summaries expanded" or "All summaries collapsed" when the "Show summaries" button is used

## Have we considered potential risks?

There are few risks as the rewrite goes to great lengths to preserve the prior version's markup.

## Images

No changes.

## Accessibility

<!-- Usually only applicable to UI changes, check the boxes if you are satisfied that your changes pass these tests -->

-   [x] [Tested with screen reader](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#screen-reader)
-   [x] [Navigable with keyboard](https://github.com/guardian/accessibility/blob/main/people-and-technology/02-physical.md#keyboard)

ctorgalson committed:
feat: converts code to an empty Drupal behavior

ctorgalson committed:
feat: rewrites main Step by Step js using theme functions

This exactly duplicates the existing behavior and markup of the existing
implementation, but makes the following improvements:

- all strings are now translateable
- nearly all HTML is themeable, including child markup for the <button>
  elements
- improved accessibility via aria-pressed and Drupal.announce() for the
  show/hide all summaries button
- much less DOM querying overall
- less code repetition

ctorgalson committed:
feat: minor refactor

- separates all-summaries toggle logic
- declares variables where used
- passes through prettier

ctorgalson committed:
feat: final variable assignment tweaks

ctorgalson committed:
feat: completes comments, aria-pressed removal

ctorgalson committed:
feat: adds aria-controls to buttons, corresponding id to summaries

ctorgalson commented:
@markconroy I know you've just gone off on holiday, but this was the PR I mentioned on #72.

tonypaulbarker commented:
@ctorgalson There was some discussion about updating the step by step pattern so I think improving the mark up we currently have rather than wholesale changes is the right way to go for this particular MR.

ctorgalson commented:
The problem is the javascript--it's currently _creating_ the unthemeable, untranslatable markup, _and_ it's much too tightly coupled to the existing twig markup.

The least impactful change in terms of existing implementations is to fix the js so the things it does can be translated and themed.

Fixing the twig markup (by e.g. outputting the buttons there) would be a great improvement, but would require a total rewrite of the js in that case too (making it much smaller and simpler on the way).

The TL;DR is the js needs to be rewritten. 

This PR is the conservative approach, but I'd love it if we opted for a major version increment. That'd allow us to greatly simplify the script and handle 100% of the markup in twig.

markconroy reviewed (approved):
I think this looks great. happy to approve it. 

I'd also like a follow-up issue to rewrite the JS and template (as a new major release perhaps) so it does _all_ the things we want. 

Great work. Thanks.

tonypaulbarker commented:
@ctorgalson to look at test failures. Looks as though behat complains about a change of label.

ctorgalson committed:
feat: handles ids for aria-controls attributes

tonypaulbarker reviewed (changes requested):
Marking as changes requested for addressing the tests

ctorgalson commented:
Fixing tests will rely on a[n upstream issue with `localgov_shared_workflows`](https://github.com/localgovdrupal/localgov_shared_workflows/issues/30) being solved.

Essentially, I need to run the tests locally to _see_ what DOM structure gets created and why that behat test complains, and I can't correctly run the tests locally with that issue outstanding (afaict, the test is wrong about the result, but I can't prove it :grin:)...

Edited to add much later: I changed the tests to dump HTML when they fail, and ran them here.

finnlewis commented:
Some merge conflicts to review

finnlewis commented:
Keen to pick this up again. 

We just merged the ESLint fixes.

Can we resolve these conflicts then incorporate these changes? 

@markconroy @tonypaulbarker perhaps we can review in the next week or two.

ctorgalson committed:
feat: ensures we don't render summary, master toggle buttons without content

ctorgalson committed:
Merge branch '2.x' into feature/2.x/72-module-javascript-needs-improvement

ctorgalson committed:
fix: adds unnecessary name for behaviour wrapping function per eslint

ctorgalson committed:
fix: replaces prettier's double quotes with single per eslint

ctorgalson committed:
fix: adds @returns to all jsdoc blocks per eslint

ctorgalson committed:
fix: updates theme function script per eslint

ctorgalson committed:
fix uses @return instead of @returns per eslint

ctorgalson committed:
fix: adds mising @param in jsdoc block per eslint

ctorgalson committed:
fix: weird conditional format fix, missing jsdoc param addition per eslint

ctorgalson committed:
fix: moves function definitions to end per eslint

ctorgalson committed:
fix: updates jsdoc comments for main behavior function per eslint

I *think* this unusual requirement is because the function actually
returns in some circumstances, so the @return is needed.

ctorgalson committed:
fix: moves toggleStepButton() to come before any calls to it

Eslint makes things much more complex.

ctorgalson committed:
feat: reorganizes entire file around contradictory eslint rules

ctorgalson committed:
fix: fix jsdocs for Drupal.theme functions.

ctorgalson committed:
fix: names Drupal.theme functions per eslint

ctorgalson committed:
fix: shortens too-long lines per eslint

ctorgalson committed:
fix: adds missing jsdoc param per eslint

ctorgalson committed:
fix: adds comma after every function param on newline per eslint

ctorgalson committed:
feat: adds debug of .step-list parent to FunctionalJavascript tests

ctorgalson committed:
feat: improves debug output, restoring original error message ouput

ctorgalson committed:
feat: corrects debug output

ctorgalson committed:
fix: corrects regression: auto-open step when current

ctorgalson commented:
@finnlewis, @markconroy, @tonypaulbarker I got the tests to pass. There was a sneaky regression in there. The tests will also now dump the `.step-list` HTML if one of the Functional Javascript tests fails.

@markconroy this includes the identical behaviour from your recent patch.

ctorgalson commented:
@msayoung I've made one major and one minor accessibility change here that it'd be good for an expert to review.

## Major: remove `aria-expanded` / `aria-pressed` from "Show summaries" button

I've removed the `aria-expanded` attribute from the "Show all summaries" button, _and_ used `Drupal.announce()` to announce "summaries expanded" or "summaries collapsed". Here's the rationale:

- following [Adrian Roselli's suggestion for this situation](https://adrianroselli.com/2024/03/check-all-expand-all-controls.html#Recap), removed `aria-expanded`:
  > Use &lt;button aria-pressed="true|false|mixed"&gt; if you want to bulk expand / collapse a bunch of disclosures.
- but because the `<button>` text _changes_, [followed MDN's advice not to use _any_ `aria` attribute to indicate state](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-pressed):
  > If we employ native HTML semantics with [<button>](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/button), we can toggle the label instead of toggling the pressed state, removing the need for the aria-pressed attribute.
- testing with Orca, *I* found the "Show summaries, Button" and "Hide summaries, Button" screenreader messages confusing because the 'hide' one is spoken immediately on expanding the button and the 'show' one is spoken immediately upon hiding them.

For that reason, I added the `Drupal.announce()` / `aria-live` messages, but I don't know if regular screenreader users would share my confusion about this interaction.

I'd be happy to rely on a static button text ("Show all summaries") and `aria-pressed` for the button, but this PR is meant to make the js themeable and translatable, without changing the UX.

## Minor: begin using `aria-controls` attributes on individual "Show summary" buttons

I've also added `aria-controls` attributes to the original buttons and `id` attributes to the step containers since the buttons and divs are not siblings in the existing markup.
